### ISSUE TITLE
Behaviour for normalization of zero-vectors

### ISSUE THREAD

**Issue by julhe** [bitshifter/glam-rs#74]
Hey all! In #71 I proposed to add ``normalized_safe()`` as a quick way to deal with vectors that might be zero (see the example). However @bitshifter said that this should be subject of discussion (on which I agree). 

So the questions are:

-  Do we need to simplify normalization of zero vectors?
-  If so, how should that function look like?

**Comment by bitshifter** [comment]
Thanks for opening the issue @julhe.

Functions that might benefit from having checked versions:
* `Vec2`, `Vec3`, `Vec4` and `Quat` `normalize`
* `Mat2`, `Mat3`, `Mat4` `inverse`

There are other functions that may depend on valid inputs, these ones in particular the input can be variable and they will divide by zero if the input is invalid.

Some math libraries always check, some never check and some provide separate methods. A quick survey found:
* Unity returns zero vector, the tolerance is not documented which is a bit crap - https://docs.unity3d.com/ScriptReference/Vector3.Normalize.html
* Unreal takes a tolerance as an input parameter and returns zero vector if the squared magnitude is outside of the tolerance - https://docs.unrealengine.com/en-US/API/Runtime/Core/Math/FVector/GetSafeNormal/index.html
* DirectXMath just checks that length is greater than zero before dividing (so the resulting vector is zero if length is zero) - https://github.com/microsoft/DirectXMath/blob/master/Inc/DirectXMathVector.inl#L9828-L9845
* Defold doesn't check - https://defold.com/ref/stable/vmath/#vmath.normalize:v1
* Godot doesn't check - https://docs.godotengine.org/en/stable/classes/class_vector3.html#class-vector3-method-normalized

This stack overflow post discusses a few other options, like flagging if the result is a degenerate case - https://stackoverflow.com/questions/722073/how-do-you-normalize-a-zero-vector

If providing checked functions there are a couple of decisions that need to be made:
* what epsilon value should be used for length? If just trying to avoid divide by zero then checking length is > 0.0 would be sufficient, but users may be interested in catching small values. If the epsilon value is greater than zero I believe it should be an input parameter
* what to do if the input is degenerate? My feeling here was a closure might offer the best flexibility for users to decide what is the appropriate behavior in this case. Some libraries return the input, others return zero and others just perform the divide by zero and return NaNs.

For example:

```
// if length is zero then F is called
fn normalize_or_else<F>(self, f: F) -> Self;
```

In the case of normalization it is not hugely complicated for users to implement their own checks so the API above may end up being as complicated as doing it yourself. Matrix inverse is a bit trickier. Calculating the determinant before inverting is wasteful with the current API, so some method of checking if the inverse succeeded could be valuable. Intially matrix invert returned an option but my benchmarking found this had a high performance impact so that wasn't a desireable solution. It would be good if any proposed solution was common between matrix invert and vec/quat normalize.

**Comment by julhe** [comment]
Okay so I revisited my code and realized that ``normalize_safe`` is actually a bad idea. It encourages you to write code that actually does nothing. Example:
```rust
// assume target == my_position
let direction = target - my_position; // -> direction == 0
let velocity = direction.normalize_safe() * delta_time; // 0 * x = 0
my_position += velocity // 0 + x = x
```
Also from my tests: a threshold of ``f32::EPSILON``, while good enough to catch NaNs, didn't prevent "garbage vectors". These lead to jerky movement in my steering movement test. (I guess this is because some vector components fluctuate between _zero_ and _something-barely-above-zero_.) So a user defined threshold should be good here.

I've then rewrote my testing code and noticed a pattern:
```rust
let direction = target - my_position; //compute some vector
let direction_length_sqr = direction.length_sqr(); //compute squared length
if direction_length_sqr > 0.01 { // check against thresshold
   let velocity = direction * (1.0 / direction_length_sqr.sqrt()); //actually compute the normalized vector
   my_position += velocity //do something
}
```
Which leads me to this! :slightly_smiling_face:
``normalize_with_threshold(self, threshold: f32) -> Option<Self> ``
```rust
let direction = target - my_position; //compute some vector
if let Some(velocity) = direction.normalize_threshold(0.01) { //check if normalizeable
   my_position += velocity; //do something
}
```
This would also allow the user to return a zero-vector if desired
```rust
let velocity = direction.normalize_thresshold(0.01).or_else(Vec3::zero());
```
Throwing that into criterion:

- just `normalize` -> 4.9008 ns
- ``normalize_with_threshold`` -> 5.9295 ns
- ``normalize_with_threshold`` (manually inline, for comparison) -> 5.9420 ns

About degenerated input: I think handling this shouldn't be our responsibility, as these problems originate on the user side and are better to be fixed there. Offering a separate ``is_degenerated()`` function for debugging/assertions (+ maybe some extra documentation on that subject) should help there. 

So, what do you think?

**Comment by bitshifter** [comment]
Hi @julhe . I'm still curious what you think about the closure method I mentioned. There are several reasons why I'm thinking about using a closure:

* Primarily because it would only require one conditional to achieve the same result as what you are doing above which involves 2 - the if inside `normalize_checked` and the `if let` in calling code
* you could do other things in the closure like log a warning, panic etc.
* The closure method would work well for other use cases like `Mat4::inverse` where returning an `Option<Mat4>` has a big performance hit, and passing an `else` value to return if below the threshold would be costly for a `Mat4`

so something like:

```rust
pub fn normalize_or_else<F: FnOnce() -> Self>(self, threshold_sq: f32, or_else: F) -> Self {
    let length_squared = self.length_squared();
    if length_squared > threshold_sq {
        self * length_squared.sqrt().recip()
    } else {
        or_else()
    }
}
```

so your example would look like:

```rust
let direction = target - my_position; //compute some vector
let velocity = direction.normalize_or_else(0.01, || Vec3::zero()); //return unit vector if above threshold or else zero
my_position += velocity; //do something
```

BTW, by degenerate input I mean something which doesn't meet the threshold.

What do you think?

**Comment by emilk** [comment]
In my experience the simplest and best way to detect whether something can be normalized is to try it and then check if the result is finite.

## Upsides

* No epsilons or thresholds that needs tweaking
* Result is guaranteed to be finite and normalized
* Very easy to explain

## Downsides

Slightly less predictable, e.g. will `Vec2(denorm, 0.0)` succeed to normalize or not?

## Example

``` rust
pub fn try_normalize(&self) -> Self {
     let normalized = self.normalize();
     if normalized.all_finite() {
         Some(normalized)
    } else {
        None
    }
}

pub fn normalize_or_zero(&self) -> Self {
     self.try_normalize().unwrap_or(Self::zero())
}
```

(using `all_finite` from https://github.com/bitshifter/glam-rs/pull/98)

**Comment by bitshifter** [comment]
Merged #127. Will close this for now, since the original issue was for vectors, I can add something similar for matrices if it's requested.
